lib/dotcom_web/components/mode_icons.ex

```diff
@@ -0,0 +1,112 @@
+defmodule DotcomWeb.Components.ModeIcons do
+  import DotcomWeb.Components.Badges
+  import DotcomWeb.Components.Svg
+  use Phoenix.Component
+
+  attr :route_name, :string, required: true
+
+  attr :type, :atom,
+    values: [
+      :subway,
+      :commuter_rail,
+      :bus,
+      :ferry,
+      :logan_express,
+      :massport_shuttle,
+      :orange_line,
+      :red_line,
+      :green_line,
+      :blue_line,
+      :mattapan_line
+    ],
+    required: true
+
+  attr :class, :string, default: ""
+
+  def mode_icon(%{type: :bus} = assigns) do
+    ~H"""
+    <.rect_icon class={"bg-yellow-500 #{@class}"}><%= @route_name %></.rect_icon>
+    """
+  end
+
+  def mode_icon(%{type: :silver_line} = assigns) do
+    ~H"""
+    <.rect_icon class={"bg-silver-500 text-white #{@class}"}><%= @route_name %></.rect_icon>
+    """
+  end
+
+  def mode_icon(%{type: :massport_shuttle} = assigns) do
+    ~H"""
+    <.circle_icon class={"bg-navy-500 text-md #{@class}"}><%= @route_name %></.circle_icon>
+    """
+  end
+
+  def mode_icon(%{type: :blue_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-blue-500 #{@class}"}>BL</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_b} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>B</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_c} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>C</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_d} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>D</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_e} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>E</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :orange_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-orange-500 #{@class}"}>OL</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :red_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-red-500 #{@class}"}>RL</.subway_icon>
+    """
+  end
```

Comment:
These could all use metro and be done programmatically.

```
def mode_icon(%{type: type} = assigns) do
  color = type |> Atom.to_string() |> String.split("_") |> List.first()
  
  ~H"""
  <.badge color={color} type="circle"><%= type_to_string(type) %></.badge>
  """
end
```

Reply:
I could see some amount of DRY benefit to this suggestion, but that would also make this code pretty opaque.

Plus, we still need a giant case statement or pattern-match for `type_to_string`.

Plus, it doesn't look like the team is aligned on this suggestion.

Reply:
You don't need a case statement or large pattern match.

```
def type_to_string(type) do
  type
  |> Atom.to_string()
  |> String.split("_")
  |> Enum.map(&String.at(&1, 0))
  |> Enum.map(&String.upcase/1)
  |> letters_to_initials()
end

def letters_to_initials([a, b, c]), do: c

def letters_to_initials(letters), do: Enum.join(letters)
```

Reply:
This just programmatically expresses what you're doing manually in the above examples.

Reply:
This program would work for 7 of the 10 highlighted clauses (there's a forgotten extra one, `:mattapan_line`, which fits this use case _but_ doesn't follow this logic, as we'd want an M instead of ML, there's no red in the name...). 

But I actually find the above examples easier to understand at a glance than this programmatic expression. 

I do agree it could stand to be a bit more concise. I think a guard expression would achieve this more easily _and_ more clearly, e.g.

```elixir
def mode_icon(%{type: type} = assigns) when type in [:red_line, :mattapan_line, :blue_line, :orange_line, ..all the greens] do
   color = tiny function that can map the color to these types
   text =  tiny function that can map the text to these types, much like cited above
  ~H"""
    <.whatever_icon class={"#{color} #{@class}"}>{text}</.thing>
  """
end
```

Reply:
Yeah, I can't edit a highlight.

Reply:
I just want to be more clear because I think people are confused. You use pattern matching to create a function for each unique case you'll encounter. We have 'bus', 'massport_shuttle', 'mattapan_line', then generic atom with three words and generic atom with two words where initials and color are in the atom.

```
def mode_icon(%{type: :bus} = assigns) do
  ~H"""
  <.badge type="square" color="yellow" class="text-yellow-800 rounded-sm">
    <%= @route_name %>
  </.badge>
  """
end

def mode_icon(%{type: :massport_shuttle} = assigns) do
 ...
end

def mode_icon(%{type: :mattapan_line} = assigns) do
 ...
end 

# other cases that don't fit the general pattern

def mode_icon(%{type: type} = assigns) do
  ~H"""
  <.badge color={type_to_color(type)} type="circle"><%= type_to_initials(type) %></.badge>
  """
end

defp type_to_color(type) do
  type
  |> Atom.to_string()
  |> String.split("_")
  |> List.first()
end

defp type_to_initials(type) do
  type
  |> Atom.to_string()
  |> String.split("_")
  |> Enum.map(&String.at(&1, 0))
  |> Enum.map(&String.upcase/1)
  |> letters_to_initials()
end

defp letters_to_initials([a, b, c]), do: c

defp letters_to_initials(letters), do: Enum.join(letters)
```

Reply:
It should take care of all cases with less, more expressive code. It eliminates the need for the other various components, as it just uses the ones already in the metro. I was hoping to communicate this without having to type out the entire module 🤷